File: lua/tst/layer-empty-tile.lua
require 'pico.check'

pico.init(true)

pico.set.dim {'!', w=480, h=480}
pico.set.show { grid=false }

-- Layer "map": 20x15 cols/rows of 16x16 tiles -> 320x240 px texture.
-- view.tile must be set at birth so '#'-mode works without pico.set.view.
do
    print("layer map: 20x15 tiles of 16x16")
    pico.layer.empty(nil, "map", {w=20, h=15}, {w=16, h=16})
    pico.set.layer("map")
    pico.set.show { grid=true }
    pico.output.clear()
    pico.set.color.draw('red')
    pico.output.draw.rect {'#', x=1, y=1, w=1, h=1, anchor='NW'}
    pico.set.color.draw('green')
    pico.output.draw.rect {'#', x=6, y=4, w=4, h=2, anchor='NW'}
    pico.set.color.draw('blue')
    pico.output.draw.rect {'#', x=20, y=15, w=1, h=1, anchor='NW'}
    pico.set.layer(nil)
end

-- Test 1: draw layer "map" 1:1 on main
do
    print("draw map 1:1")
    pico.output.clear()
    pico.output.draw.layer("map", {'!', x=0, y=0, w=320, h=240, anchor='NW'})
    pico.check("layer-empty-tile-01")
end

-- Test 2: draw layer "map" scaled up
do
    print("draw map scaled")
    pico.output.clear()
    pico.output.draw.layer("map", {'!', x=0, y=0, w=480, h=360, anchor='NW'})
    pico.check("layer-empty-tile-02")
end

-- Layer "fx": no tile arg -> dim is pixels, view.tile stays {0,0}
do
    print("layer fx: plain pixel layer (no tile)")
    pico.layer.empty(nil, "fx", {w=64, h=64})
    pico.set.layer("fx")
    pico.output.clear()
    pico.set.color.draw('yellow')
    pico.output.draw.rect {'!', x=8, y=8, w=48, h=48, anchor='NW'}
    pico.set.layer(nil)
end

-- Test 3: draw both layers
do
    print("draw map + fx")
    pico.output.clear()
    pico.output.draw.layer("map", {'!', x=0,   y=0, w=320, h=240, anchor='NW'})
    pico.output.draw.layer("fx",  {'!', x=340, y=0, w=128, h=128, anchor='NW'})
    pico.check("layer-empty-tile-03")
end

pico.init(false)

File: tst/layer-empty-tile.c
#include "pico.h"
#include "../check.h"

int main (void) {
    pico_init(1);

    Pico_Rel_Dim phy = { '!', {480, 480}, NULL };
    pico_set_dim(&phy);
    pico_set_effect_grid("root", 0);

    // Layer "map": 20x15 cols/rows of 16x16 tiles -> 320x240 px texture.
    // view.tile must be set at birth so '#'-mode works without pico_set_scene.
    {
        puts("layer map: 20x15 tiles of 16x16");
        Pico_Abs_Dim tile = { 16, 16 };
        pico_layer_empty(NULL, "map", (Pico_Abs_Dim){20, 15}, &tile);
        pico_set_effect_grid("map", 1);
        pico_output_clear("map");
        pico_set_pencil_color("map", PICO_COLOR_RED);
        pico_output_draw_rect("map", &(Pico_Rel_Rect){ '#', {1, 1, 1, 1}, PICO_ANCHOR_NW, NULL });
        pico_set_pencil_color("map", PICO_COLOR_GREEN);
        pico_output_draw_rect("map", &(Pico_Rel_Rect){ '#', {6, 4, 4, 2}, PICO_ANCHOR_NW, NULL });
        pico_set_pencil_color("map", PICO_COLOR_BLUE);
        pico_output_draw_rect("map", &(Pico_Rel_Rect){ '#', {20, 15, 1, 1}, PICO_ANCHOR_NW, NULL });
    }

    // Test 1: draw layer "map" 1:1 on main, confirm tile-aligned content
    {
        puts("draw map 1:1");
        pico_output_clear("root");
        pico_output_draw_layer("root", "map", &(Pico_Rel_Rect){ '!', {0, 0, 320, 240}, PICO_ANCHOR_NW, NULL });
        _pico_check("layer-empty-tile-01");
    }

    // Test 2: draw layer "map" scaled up, grid still aligned visually
    {
        puts("draw map scaled");
        pico_output_clear("root");
        pico_output_draw_layer("root", "map", &(Pico_Rel_Rect){ '!', {0, 0, 480, 360}, PICO_ANCHOR_NW, NULL });
        _pico_check("layer-empty-tile-02");
    }

    // Layer "fx": tile == NULL -> dim is pixels, view.tile stays {0,0}
    {
        puts("layer fx: plain pixel layer (tile=NULL)");
        pico_layer_empty(NULL, "fx", (Pico_Abs_Dim){64, 64}, NULL);
        pico_output_clear("fx");
        pico_set_pencil_color("fx", PICO_COLOR_YELLOW);
        pico_output_draw_rect("fx", &(Pico_Rel_Rect){ '!', {8, 8, 48, 48}, PICO_ANCHOR_NW, NULL });
    }

    // Test 3: draw both layers, confirm tile-mode and pixel-mode coexist
    {
        puts("draw map + fx");
        pico_output_clear("root");
        pico_output_draw_layer("root", "map", &(Pico_Rel_Rect){ '!', {0,   0, 320, 240}, PICO_ANCHOR_NW, NULL });
        pico_output_draw_layer("root", "fx",  &(Pico_Rel_Rect){ '!', {340, 0, 128, 128}, PICO_ANCHOR_NW, NULL });
        _pico_check("layer-empty-tile-03");
    }

    pico_init(0);
    return 0;
}
